pkg/otlp/logs.go

```diff
@@ -0,0 +1,314 @@
+	if err != nil {
+		log.Errorf("ProcessLogIngest: failed to get data to unmarshal: %v", err)
+		setFailureResponse(ctx, fasthttp.StatusBadRequest, err.Error())
+		return
+	}
+
+	request, err := unmarshalLogRequest(data)
+	if err != nil {
+		log.Errorf("ProcessLogIngest: failed to unpack Data: %s with err %v", string(data), err)
+		setFailureResponse(ctx, fasthttp.StatusBadRequest, "Unable to unmarshal traces")
+		return
+	}
+
+	numTotalRecords, numFailedRecords := ingestLogs(request, myid)
+	usageStats.UpdateStats(uint64(len(data)), uint64(max(0, numTotalRecords-numFailedRecords)), myid)
+
+	// Send the appropriate response.
+	setLogIngestionResponse(ctx, numTotalRecords, numFailedRecords)
+}
+
+func ingestLogs(request *collogpb.ExportLogsServiceRequest, myid int64) (int, int) {
+	now := utils.GetCurrentTimeInMs()
+	timestampKey := config.GetTimeStampKey()
+	var jsParsingStackbuf [utils.UnescapeStackBufSize]byte
+	localIndexMap := make(map[string]string)
+	idxToStreamIdCache := make(map[string]string)
+	cnameCacheByteHashToStr := make(map[uint64]string)
+	pleArray := make([]*segwriter.ParsedLogEvent, 0)
+	numTotalRecords := 0
+	numFailedRecords := 0
+
+	for _, resourceLog := range request.ResourceLogs {
+		resource, indexName, err := extractResourceInfo(resourceLog)
+		if err != nil {
+			log.Errorf("ingestLogs: failed to extract resource info: %v", err)
+			for _, scopeLog := range resourceLog.ScopeLogs {
+				numTotalRecords += len(scopeLog.LogRecords)
+				numFailedRecords += len(scopeLog.LogRecords)
+			}
+
+			continue
+		}
+
+		for _, scopeLog := range resourceLog.ScopeLogs {
+			scope, err := extractScopeInfo(scopeLog)
+			if err != nil {
+				log.Errorf("ingestLogs: failed to extract scope info: %v", err)
+				numTotalRecords += len(scopeLog.LogRecords)
+				numFailedRecords += len(scopeLog.LogRecords)
+
+				continue
+			}
+
+			for _, logRecord := range scopeLog.LogRecords {
+				numTotalRecords++
+				record, err := extractLogRecord(logRecord, resource, scope)
+				if err != nil {
+					log.Errorf("ingestLogs: failed to extract log record: %v", err)
+					numFailedRecords++
+					continue
+				}
+
+				jsonBytes, err := json.Marshal(record)
+				if err != nil {
+					log.Errorf("ingestLogs: failed to marshal log record; err=%v", err)
+					numFailedRecords++
+					continue
+				}
+
+				ple, err := segwriter.GetNewPLE(jsonBytes, now, indexName, &timestampKey, jsParsingStackbuf[:])
+				if err != nil {
+					log.Errorf("ingestLogs: failed to get new PLE, jsonBytes: %v, err: %v", jsonBytes, err)
+					numFailedRecords++
+					continue
+				}
+
+				if timestampMs := record.TimeUnixNano / 1_000_000; timestampMs > 0 {
+					ple.SetTimestamp(timestampMs)
+				}
+
+				pleArray = append(pleArray, ple)
+			}
+		}
+
+		shouldFlush := false
+		err = writer.ProcessIndexRequestPle(now, indexName, shouldFlush, localIndexMap, myid, 0, idxToStreamIdCache, cnameCacheByteHashToStr, jsParsingStackbuf[:], pleArray)
```

Comment:
some point in future, we can expose a method from writer that will take the request json object and it can write to wip directly and avoid this marshall/unmarshall. Same concept we will apply when we support ingesting via csv file directly.